### ISSUE TITLE
[webview] WebView not parsing and rendering contents until visible

### ISSUE THREAD

**Issue by jmgeffroy** [flutter/flutter#97483]
Hi, this is a follow-up to [another issue](https://github.com/flutter/flutter/issues/95938), discussed with @darshankawar. This issue was not clearly spotted, so I took more time on a small test project. I now have more precisely qualified the issue.

The sample project is here: https://github.com/Mantano/pagination_tests.
There are 2 branches to test:
- https://github.com/Mantano/pagination_tests/tree/simplified: this one uses webview_flutter
- https://github.com/Mantano/pagination_tests/tree/flutter_inappwebview: this one uses flutter_inappwebview

In my sample project, 6 webviews are placed in a [PreloadPageView](https://pub.dev/packages/preload_page_view). This PreloadPageView allows defining how many pages are preloaded (`preloadPagesCount` param).

[When using webview_flutter](https://github.com/Mantano/pagination_tests/tree/simplified), I see this, in the chrome devtools:

<img width="746" alt="Capture d’écran 2022-01-29 à 17 37 15" src="https://user-images.githubusercontent.com/4281439/151671304-5d4a6c43-f2dd-4ced-ac3b-352bcc199a8a.png">

And when looking at one of the "empty - never attached" webviews:

<img width="995" alt="Capture d’écran 2022-01-29 à 17 37 27" src="https://user-images.githubusercontent.com/4281439/151671320-0f7ae435-35a6-417c-be31-0a8a1726edd4.png">

As you can see, the document has been parsed, but not pre-rendered ("empty - never attached"). As a consequence, if I add some JS code that does some computation based on the rendered contents, it will fail (as it does in our real ebook reader app).

[When using  flutter_inappwebview](https://github.com/Mantano/pagination_tests/tree/flutter_inappwebview), I see this, which is what I expect:

<img width="724" alt="Capture d’écran 2022-01-29 à 17 19 40" src="https://user-images.githubusercontent.com/4281439/151671357-fffa9d77-614b-4749-a053-56991fb346c6.png">

And looking at one of the not visible webviews:

<img width="645" alt="Capture d’écran 2022-01-29 à 17 20 14" src="https://user-images.githubusercontent.com/4281439/151671366-54e358b2-07cc-484e-8d93-63294ddb96fa.png">

Here, it's OK and any JS can do computations using the rendered contents dimensions.

**In this flutter_inappwebview based example, there is something very strange in the devtools screenshot: all webviews appear to be at the same position (0,0)**. And it's not a problem since swiping in the pageview works perfectly. _Maybe flutter_inappwebview doesn't place the native webview at their logical position in the UI, but at (0,0) and it allows to get the prerenderings? Just a guess..._
I'm pretty sure (100%) that this isn't a behavior induced by PreloadPageView, since the related code is exactly the same between the 2 branches.

I hope these explanations are sufficiently precise. I'm pretty sure there is a simple fix. I'd like to continue using webview_flutter instead of flutter_inappwebview for many reasons...

**EDIT** Please note that I am using Hybrid Composition on Android since it provides better performances, and it matches iOS where this is the only option available. But the issue doesn't occur on Android when using Virtual Display. Maybe this clue can also help? 

Here is my doctor -v output:

```
[✓] Flutter (Channel dev, 2.10.0-0.3.pre, on macOS 12.0.1 21A559 darwin-arm, locale fr-FR)
    • Flutter version 2.10.0-0.3.pre at /Applications/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fdd0af78bb (il y a 4 jours), 2022-01-25 22:01:33 -0600
    • Engine revision 5ac30ef0c7
    • Dart version 2.16.0 (build 2.16.0-134.5.beta)
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Applications/android-sdk-macosx
    • Platform android-32, build-tools 31.0.0
    • ANDROID_HOME = /Applications/android-sdk-macosx
    • Java binary at: /Library/Java/JavaVirtualMachines/jdk-11.0.7.jdk/Contents/Home/bin/java
    • Java version Java(TM) SE Runtime Environment 18.9 (build 11.0.7+8-LTS)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode-13.2.1.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] Android Studio (version 2020.3)
    • Android Studio at /Users/jmgeffroy/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7784292/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] Android Studio (version 2020.3)
    • Android Studio at /Users/jmgeffroy/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7678000/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[☠] IntelliJ IDEA Ultimate Edition (the doctor check crashed)
    ✗ Due to an error, the doctor check did not complete. If the error message below is not helpful, please let us know about this issue at https://github.com/flutter/flutter/issues.
    ✗ FormatException: Unexpected extension byte (at offset 5)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.1)
    • IntelliJ at /Users/jmgeffroy/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/213.6461.79/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3)
    • IntelliJ at /Users/jmgeffroy/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/213.5744.223/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 12.0.1 21A559 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 97.0.4692.99

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

```

**Comment by darshankawar** [comment]
@jmgeffroy 
Can you please check if [this](https://github.com/flutter/flutter/issues/51875) issue sounds similar / address the issue you described ?

**Comment by jmgeffroy** [comment]
Hello, @darshankawar thank you for your fast response. Yes, definitely. It looks very similar! 
BTW: when we developed our ebook reading apps "natively" on Android and iOS, in Kotlin and Swift, we used to do a lot of offscreen prerendering. 
Remember that flutter_inappwebview implementation does prerender. It even provides a headless webview, that allows doing server-side rendering, for example. Sorry to talk about another webview, but I thought this might help. It could be worth looking at it or contacting the developers?

**Comment by jmgeffroy** [comment]
Aa I mentioned, an important clue is probably the (0,0) position of ALL the webviews that are in the PageView, when I use flutter_inappwebview. You can see that in my screenshots of devtools in my first comment. With webview_flutter, you can see that the native webviews do have positions that match their position inside the PageView: only one of them, the visible one, is at (0,0).

**Comment by darshankawar** [comment]
@jmgeffroy 
Thanks for further details. I believe you should follow-up in the original issue for further updates.
I am going ahead and closing this as a duplicate of https://github.com/flutter/flutter/issues/51875.

**Comment by jmgeffroy** [comment]
Hi @darshankawar Thanks a lot, I'll follow! Good luck with your investigations.

**Comment by jmgeffroy** [comment]
Last update: I tried again my "scratch project", with flutter_inappwebview. And I have to correct one error I wrote earlier: I had NOT activated Hybrid Composition with this webview. When activating it, I have the same "empty, never attached") indication in devtools for the offscreen webviews in my PageView. Sorry for misleading you. 
So the only difference in my tests between webview_flutter and flutter_inappwebview with virtual displays is simply the position displayed by the devtools, but this is not at all a problem, nor a clue.

So this issue can be closed, and there is no relevant info in my posts from flutter_inappwebview...
